### DISCUSSION TITLE
Can egg be used to write compilers ?

### DISCUSSION THREAD

**Discussion by pcarbonn** [egraphs-good/egg#118] [General]
The paper presents 'compiler optimizations' and 'program synthesis' as target applications of egg: indeed, such applications need to consider the full set of equivalent formulations of a program.

But what about simple compilers?  Eliminating the 'phase ordering' problem is nice, but isn't it at the cost of compilation speed?  After all, we don't need the full set of equivalent programs: more complex formulations can be pruned.  Are there special techniques to speed up egg when building compilers?

**Comment by mwillsey** [comment]
Sure! You could do this in principle, but you are right that performance would a chief concern. We are still working on building a "real" compiler using egg, so we don't exactly know what the necessary techniques are. Managing the search space size would undoubtedly be one of those techniques.

**Reply by zellyn** [reply]
Just out of curiousity: are you actively working on using egg to build a compiler, and how far along are you?

**Reply by mwillsey** [reply]
I am not currently using it to build a programming language compiler, but I hope to soon (after finishing some other projects).
